torch/distributed/_shard/checkpoint/state_dict_loader.py

```diff
@@ -132,25 +139,56 @@ def load_state_dict(
         >>> # ensure correct behavior.
         >>> my_model.load_state_dict(model_state_dict)
     """
-
-    metadata = storage_reader.read_metadata()
-    bytes_read_requests, tensor_read_requests = _reshard_and_prepare_read_request(
-        state_dict=state_dict, metadata_from_storage=metadata
-    )
-    bytes_futures = storage_reader.read_bytes(bytes_read_requests)
-    tensor_futures = storage_reader.read_tensors(tensor_read_requests)
-
-    bytes_futures.wait()
-
-    # Addtional steps are required to convert the bytes to its original type
-    # Note that this is NOT inplace,
-    # it creating a new object and replace what's in the state dict
-    for req in bytes_read_requests:
-        # Ensure the BytesIO is rewound
-        req.bytes.seek(0)
-        state_dict[req.fqn] = torch.load(req.bytes)
-
-    tensor_futures.wait()
+    is_coordinator = not dist.is_initialized() or dist.get_rank(process_group) == coordinator_rank
```

Comment:
I think we should just enforce that `dist.is_initialized()` and throw an error otherwise.

Reply:
This methods must work without dist to support the local inference case.

Reply:
So I'm worried about the following case, where the user forgets to initialize the process group (but intended to use a distributed checkpoint and not just local). Then this API would kinda silently do the incorrect thing which makes things harder to debug.

Can we maybe have an additional parameter to the API (ex: `local` or `no_dist`) that explicitly communicates this intention from the user? Or we could have a separate API like `load_state_dict_local` to do something like this. I feel relying on `dist.initialized()` to determine whether it is the local inference case or a distributed checkpoint is a bit fragile.

Reply:
I guess we can add it as an experimental argument and collect feedback from users.

I think, in general, the scenario you describe would only happen if the user intends to load the checkpoint in a SPMD fashion but without using any of the existing distributed modules.

You can't have a ShardedTensor in a state_dict without initing a PG first - ShardedTensor simply fails to construct without one.

Reply:
Oh just found out we already have a conversation on this haha, have a similar concern, commented in https://github.com/pytorch/pytorch/pull/77091#discussion_r875301968 My general feeling is that if we could have a way that user don't need to aware of/or no explicit flagging that might be better as once we have this flag in the API, it's hard to remove it later, let me know if the approach I described make sense or not :)